File: golang/internal/application/usecase/order/new_order_usecase.go
package order

import (
	"errors"

	"folio/codinginterview/internal/application/repository"
	"folio/codinginterview/internal/domain"

	"github.com/shopspring/decimal"
)

var (
	ErrNewOrderUserAlreadyExists = errors.New("user already has account")
	ErrNewOrderAmountTooSmall    = errors.New("amount is too small")
)

type NewOrderUsecaseInput struct {
	UserId domain.UserId
	Amount decimal.Decimal
}

type NewOrderUsecase struct {
	accountRepo   repository.AccountRepository
	portfolioRepo repository.PortfolioRepository
}

func NewNewOrderUsecase(
	accountRepo repository.AccountRepository,
	portfolioRepo repository.PortfolioRepository,
) *NewOrderUsecase {
	return &NewOrderUsecase{
		accountRepo:   accountRepo,
		portfolioRepo: portfolioRepo,
	}
}

func (u *NewOrderUsecase) Run(input NewOrderUsecaseInput) error {
	if input.Amount.LessThan(domain.MinOperationAmount) {
		return ErrNewOrderAmountTooSmall
	}

	exists, err := u.accountRepo.Exists(input.UserId)
	if err != nil {
		return err
	}
	if exists {
		return ErrNewOrderUserAlreadyExists
	}

	portfolio, err := u.portfolioRepo.Get()
	if err != nil {
		return err
	}

	account := domain.OpenAccount(input.Amount, portfolio)

	return u.accountRepo.Upsert(input.UserId, account)
}

File: typescript/src/application/usecase/order/newOrderUsecase.ts
import Decimal from "decimal.js";
import { AccountRepository } from "../../repository/accountRepository.js";
import { PortfolioRepository } from "../../repository/portfolioRepository.js";
import { Account } from "../../../domain/account.js";
import { AppConstants } from "../../../domain/appConstants.js";
import { UserId } from "../../../domain/userId.js";

export interface NewOrderUsecaseInput {
  userId: UserId;
  amount: Decimal;
}

export class NewOrderUsecaseException extends Error {}
export class NewOrderUserAlreadyExistsException extends NewOrderUsecaseException {
  constructor() {
    super("user already exists");
  }
}
export class NewOrderAmountTooSmallException extends NewOrderUsecaseException {
  constructor() {
    super("amount too small");
  }
}

export class NewOrderUsecase {
  constructor(
    private readonly accountRepository: AccountRepository,
    private readonly portfolioRepository: PortfolioRepository,
  ) {}

  async run(input: NewOrderUsecaseInput): Promise<void> {
    if (input.amount.lessThan(AppConstants.minOperationAmount)) {
      throw new NewOrderAmountTooSmallException();
    }
    const exists = await this.accountRepository.exists(input.userId);
    if (exists) {
      throw new NewOrderUserAlreadyExistsException();
    }
    const portfolio = await this.portfolioRepository.get();
    const account = Account.openAccount(input.amount, portfolio);
    await this.accountRepository.upsert(input.userId, account);
  }
}
